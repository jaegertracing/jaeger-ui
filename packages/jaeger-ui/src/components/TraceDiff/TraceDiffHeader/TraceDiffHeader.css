/*
Copyright (c) 2018 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.TraceDiffHeader {
  background: #fff;
  border-bottom: 1px solid #e8e8e8;
  box-shadow: 0 1px 4px rgba(0, 0, 0, 0.1);
  display: flex;
  flex: 0;
  position: relative;
  z-index: 1;
  align-items: stretch;
}

.TraceDiffHeader--labelItem {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 50px;
  background: #f5f5f5;
  border-right: 1px solid #e8e8e8;
}

.TraceDiffHeader--label {
  margin: 0;
  font-size: 24px;
  font-weight: 600;
  color: #333;
}

.TraceDiffHeader--traceSection {
  display: flex;
  flex: 1;
  min-width: 0;
  cursor: pointer;
  transition: background-color 0.2s ease;
}

.TraceDiffHeader--traceSection:hover {
  background-color: #f9f9f9;
}

.TraceDiffHeader--divider {
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 0 15px;
  background: #f5f5f5;
  border-right: 1px solid #e8e8e8;
  border-left: 1px solid #e8e8e8;
}

.TraceDiffHeader--vsContainer {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 36px;
  height: 36px;
  border-radius: 50%;
  background: #199;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.15);
  transition:
    transform 0.2s ease,
    box-shadow 0.2s ease;
}

.TraceDiffHeader--vsContainer:hover {
  transform: scale(1.05);
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.2);
}

.TraceDiffHeader--vsLabel {
  font-size: 14px;
  font-weight: 700;
  color: white;
  letter-spacing: 1px;
  text-shadow: 0 1px 1px rgba(0, 0, 0, 0.2);
}

.TraceDiffHeader--popover .ant-popover-title {
  padding: 10px;
  border-bottom: 1px solid #e8e8e8;
}

.TraceDiffHeader--popover .ant-popover-inner-content {
  padding: 0;
}

.TraceDiffHeader--popover .ant-popover-inner {
  border-radius: 4px;
  box-shadow:
    0 3px 6px -4px rgba(0, 0, 0, 0.12),
    0 6px 16px 0 rgba(0, 0, 0, 0.08),
    0 9px 28px 8px rgba(0, 0, 0, 0.05);
}
