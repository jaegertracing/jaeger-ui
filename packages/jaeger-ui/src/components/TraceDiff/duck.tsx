// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { Action, ActionFunctionAny, createActions, handleActions } from 'redux-actions';

import TTraceDiffState from '../../types/TTraceDiffState';
import generateActionTypes from '../../utils/generate-action-types';
import guardReducer from '../../utils/guardReducer';

// payloads
type TTraceIdValue = { traceID: string };
type TNewStateValue = { newState: TTraceDiffState };
type TTraceDiffActions = {
  [actionName: string]: ActionFunctionAny<Action<TTraceIdValue | TNewStateValue>>;
};

export function newInitialState(): TTraceDiffState {
  return {
    cohort: [],
    a: null,
    b: null,
  };
}

export const actionTypes = generateActionTypes('@jaeger-ui/trace-diff', [
  'COHORT_ADD_TRACE',
  'COHORT_REMOVE_TRACE',
  'DIFF_SET_A',
  'DIFF_SET_B',
  'FORCE_STATE',
]);

const fullActions = createActions<TTraceIdValue | TNewStateValue>({
  [actionTypes.COHORT_ADD_TRACE]: (traceID: string) => ({ traceID }),
  [actionTypes.COHORT_REMOVE_TRACE]: (traceID: string) => ({ traceID }),
  [actionTypes.DIFF_SET_A]: (traceID: string) => ({ traceID }),
  [actionTypes.DIFF_SET_B]: (traceID: string) => ({ traceID }),
  [actionTypes.FORCE_STATE]: (newState: TTraceDiffState) => ({ newState }),
});

export const actions = (fullActions as unknown as { jaegerUi: { traceDiff: TTraceDiffActions } }).jaegerUi
  .traceDiff;

function cohortAddTrace(state: TTraceDiffState, { traceID }: TTraceIdValue) {
  if (state.cohort.indexOf(traceID) >= 0) {
    return state;
  }
  const cohort = state.cohort.slice();
  cohort.push(traceID);
  return { ...state, cohort };
}

function cohortRemoveTrace(state: TTraceDiffState, { traceID }: TTraceIdValue) {
  const i = state.cohort.indexOf(traceID);
  if (i < 0) {
    return state;
  }
  const cohort = state.cohort.slice();
  cohort.splice(i, 1);
  const a = state.a === traceID ? null : state.a;
  const b = state.b === traceID ? null : state.b;
  return { ...state, a, b, cohort };
}

function diffSetA(state: TTraceDiffState, { traceID }: TTraceIdValue): TTraceDiffState {
  return { ...state, a: traceID };
}

function diffSetB(state: TTraceDiffState, { traceID }: TTraceIdValue): TTraceDiffState {
  return { ...state, b: traceID };
}

function forceState(state: TTraceDiffState, { newState }: TNewStateValue) {
  return newState;
}

export default handleActions<TTraceDiffState>(
  {
    [actionTypes.COHORT_ADD_TRACE]: guardReducer(cohortAddTrace),
    [actionTypes.COHORT_REMOVE_TRACE]: guardReducer(cohortRemoveTrace),
    [actionTypes.DIFF_SET_A]: guardReducer(diffSetA),
    [actionTypes.DIFF_SET_B]: guardReducer(diffSetB),
    [actionTypes.FORCE_STATE]: guardReducer(forceState),
  },
  newInitialState()
);
