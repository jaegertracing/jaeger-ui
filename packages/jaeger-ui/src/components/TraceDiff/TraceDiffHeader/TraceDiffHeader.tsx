// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Popover } from 'antd';

import CohortTable from './CohortTable';
import TraceHeader from './TraceHeader';
import TraceIdInput from './TraceIdInput';

import { FetchedTrace, TNil } from '../../../types';

import './TraceDiffHeader.css';

type Props = {
  a: FetchedTrace | TNil;
  b: FetchedTrace | TNil;
  cohort: FetchedTrace[];
  diffSetA: (traceId: string) => void;
  diffSetB: (traceId: string) => void;
};

type State = {
  tableVisible: ('a' | 'b') | null;
};

export default class TraceDiffHeader extends React.PureComponent<Props, State> {
  _toggleTableA: (showTable: boolean) => void;
  _toggleTableB: (showTable: boolean) => void;
  _diffSetA: (traceID: string) => void;
  _diffSetB: (traceID: string) => void;

  state = {
    tableVisible: null,
  };

  constructor(props: Props) {
    super(props);
    this._toggleTableA = this._toggleTable.bind(this, 'a');
    this._toggleTableB = this._toggleTable.bind(this, 'b');
    this._diffSetA = this._diffSetTrace.bind(this, 'a');
    this._diffSetB = this._diffSetTrace.bind(this, 'b');
  }

  _toggleTable(which: 'a' | 'b', visible: boolean) {
    const tableVisible = visible ? which : null;
    this.setState({ tableVisible });
  }

  _diffSetTrace(which: 'a' | 'b', id: string) {
    if (which === 'a') {
      this.props.diffSetA(id);
    } else {
      this.props.diffSetB(id);
    }
    this.setState({ tableVisible: null });
  }

  render() {
    const { a, b, cohort } = this.props;
    const { tableVisible } = this.state;
    const {
      data: aData = undefined,
      id: aId = undefined,
      state: aState = undefined,
      error: aError = undefined,
    } = a || {};
    const {
      data: bData = undefined,
      id: bId = undefined,
      state: bState = undefined,
      error: bError = undefined,
    } = b || {};
    const selection: Record<string, { label: 'A' | 'B' }> = {};
    if (aId) selection[aId] = { label: 'A' };
    if (bId) selection[bId] = { label: 'B' };
    const cohortTableA = (
      <CohortTable cohort={cohort} current={aId} selectTrace={this._diffSetA} selection={selection} />
    );
    const cohortTableB = (
      <CohortTable cohort={cohort} current={bId} selectTrace={this._diffSetB} selection={selection} />
    );
    return (
      <header className="TraecDiffHeader">
        <div className="TraecDiffHeader--labelItem-darkened">
          <h1 className="ub-m0">A</h1>
        </div>
        <Popover
          overlayClassName="TraceDiffHeader--popover"
          trigger="click"
          placement="bottomLeft"
          title={<TraceIdInput selectTrace={this._diffSetA} />}
          content={cohortTableA}
          open={tableVisible === 'a'}
          onOpenChange={this._toggleTableA}
        >
          <div className="ub-flex u-flex-1">
            <TraceHeader
              duration={aData && aData.duration}
              error={aError}
              startTime={aData && aData.startTime}
              state={aState}
              totalSpans={aData && aData.spans && aData.spans.length}
              traceID={aId}
              traceName={aData && aData.traceName}
            />
          </div>
        </Popover>
        <div className="TraecDiffHeader--labelItem">
          <h2 className="u-tx-muted ub-mx3 ub-my0">VS</h2>
        </div>
        <div className="TraecDiffHeader--labelItem-darkened">
          <h1 className="ub-m0">B</h1>
        </div>
        <Popover
          overlayClassName="TraceDiffHeader--popover"
          trigger="click"
          placement="bottomLeft"
          title={<TraceIdInput selectTrace={this._diffSetB} />}
          content={cohortTableB}
          open={tableVisible === 'b'}
          onOpenChange={this._toggleTableB}
        >
          <div className="ub-flex u-flex-1">
            <TraceHeader
              duration={bData && bData.duration}
              error={bError}
              startTime={bData && bData.startTime}
              state={bState}
              totalSpans={bData && bData.spans && bData.spans.length}
              traceID={bId}
              traceName={bData && bData.traceName}
            />
          </div>
        </Popover>
      </header>
    );
  }
}
