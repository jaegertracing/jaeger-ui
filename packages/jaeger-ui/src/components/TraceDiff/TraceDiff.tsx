// Copyright (c) 2018 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { History as RouterHistory } from 'history';
import queryString from 'query-string';
import { connect } from 'react-redux';
import { match } from 'react-router-dom';
import { bindActionCreators, Dispatch } from 'redux';

import { actions as diffActions } from './duck';
import { getUrl, TDiffRouteParams } from './url';
import TraceDiffGraph from './TraceDiffGraph';
import TraceDiffHeader from './TraceDiffHeader';
import * as jaegerApiActions from '../../actions/jaeger-api';
import { TOP_NAV_HEIGHT } from '../../constants';
import { FetchedTrace, TNil, ReduxState } from '../../types';
import TTraceDiffState from '../../types/TTraceDiffState';
import pluckTruthy from '../../utils/ts/pluckTruthy';

import './TraceDiff.css';

type TStateProps = {
  a: string | undefined;
  b: string | undefined;
  cohort: string[];
  tracesData: Map<string, FetchedTrace>;
  traceDiffState: TTraceDiffState;
};

type TDispatchProps = {
  fetchMultipleTraces: (ids: string[]) => void;
  forceState: (state: TTraceDiffState) => void;
};

type TOwnProps = {
  history: RouterHistory;
  match: match<TDiffRouteParams>;
};

type TState = {
  graphTopOffset: number;
};

function syncStates(
  urlValues: TTraceDiffState,
  reduxValues: TTraceDiffState,
  forceState: (newState: TTraceDiffState) => void
) {
  const { a: urlA, b: urlB } = urlValues;
  const { a: reduxA, b: reduxB } = reduxValues;
  if (urlA !== reduxA || urlB !== reduxB) {
    forceState(urlValues);
    return;
  }
  const urlCohort = new Set(urlValues.cohort);
  const reduxCohort = new Set(reduxValues.cohort || []);
  if (urlCohort.size !== reduxCohort.size) {
    forceState(urlValues);
    return;
  }
  const needSync = Array.from(urlCohort).some(id => !reduxCohort.has(id));
  if (needSync) {
    forceState(urlValues);
  }
}

// export class TraceDiffImpl extends React.PureComponent<Props, State> {
export class TraceDiffImpl extends React.PureComponent<TStateProps & TDispatchProps & TOwnProps, TState> {
  state = {
    graphTopOffset: TOP_NAV_HEIGHT,
  };

  headerWrapperElm: HTMLDivElement | TNil = null;

  componentDidMount() {
    this.processProps();
  }

  componentDidUpdate() {
    this.setGraphTopOffset();
    this.processProps();
  }

  headerWrapperRef = (elm: HTMLDivElement | TNil) => {
    this.headerWrapperElm = elm;
    this.setGraphTopOffset();
  };

  setGraphTopOffset() {
    if (this.headerWrapperElm) {
      const graphTopOffset = TOP_NAV_HEIGHT + this.headerWrapperElm.clientHeight;
      if (this.state.graphTopOffset !== graphTopOffset) {
        this.setState({ graphTopOffset });
      }
    } else {
      this.setState({ graphTopOffset: TOP_NAV_HEIGHT });
    }
  }

  processProps() {
    const { a, b, cohort, fetchMultipleTraces, forceState, tracesData, traceDiffState } = this.props;
    syncStates({ a, b, cohort }, traceDiffState, forceState);
    const cohortData = cohort.map(id => tracesData.get(id) || { id, state: null });
    const needForDiffs = cohortData.filter(ft => ft.state == null).map(ft => ft.id);
    if (needForDiffs.length) {
      fetchMultipleTraces(needForDiffs);
    }
  }

  diffSetUrl(change: { newA?: string | TNil; newB?: string | TNil }) {
    const { newA, newB } = change;
    const { a, b, cohort, history } = this.props;
    const url = getUrl({ a: newA || a, b: newB || b, cohort });
    history.push(url);
  }

  diffSetA = (id: string) => {
    const newA = id.toLowerCase();
    this.diffSetUrl({ newA });
  };

  diffSetB = (id: string) => {
    const newB = id.toLowerCase();
    this.diffSetUrl({ newB });
  };

  render() {
    const { a, b, cohort, tracesData } = this.props;
    const { graphTopOffset } = this.state;
    const traceA = a ? tracesData.get(a) || { id: a } : null;
    const traceB = b ? tracesData.get(b) || { id: b } : null;
    const cohortData: FetchedTrace[] = cohort.map(id => tracesData.get(id) || { id });
    return (
      <React.Fragment>
        <div key="header" ref={this.headerWrapperRef}>
          <TraceDiffHeader
            key="header"
            a={traceA}
            b={traceB}
            cohort={cohortData}
            diffSetA={this.diffSetA}
            diffSetB={this.diffSetB}
          />
        </div>
        <div key="graph" className="TraceDiff--graphWrapper" style={{ top: graphTopOffset }}>
          <TraceDiffGraph a={traceA} b={traceB} />
        </div>
      </React.Fragment>
    );
  }
}

// TODO(joe): simplify but do not invalidate the URL
export function mapStateToProps(state: ReduxState, ownProps: { match: match<TDiffRouteParams> }) {
  const { a, b } = ownProps.match.params;
  const { cohort: origCohort = [] } = queryString.parse(state.router.location.search);
  const fullCohortSet: Set<string> = new Set(pluckTruthy([a, b].concat(origCohort)));
  const cohort: string[] = Array.from(fullCohortSet);
  const { traces } = state.trace;
  const kvPairs = cohort.map<[string, FetchedTrace]>(id => [id, traces[id] || { id, state: null }]);
  const tracesData: Map<string, FetchedTrace> = new Map(kvPairs);
  return {
    a,
    b,
    cohort,
    tracesData,
    traceDiffState: state.traceDiff,
  };
}

// export for tests
export function mapDispatchToProps(dispatch: Dispatch<any>) {
  const { fetchMultipleTraces } = bindActionCreators(jaegerApiActions, dispatch);
  const { forceState } = bindActionCreators(diffActions, dispatch);
  return { fetchMultipleTraces, forceState };
}

export default connect<TStateProps, TDispatchProps, TOwnProps, ReduxState>(
  mapStateToProps,
  mapDispatchToProps
)(TraceDiffImpl);
