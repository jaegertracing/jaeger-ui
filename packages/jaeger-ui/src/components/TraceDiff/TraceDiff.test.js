// Copyright (c) 2019 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { shallow } from 'enzyme';
import queryString from 'query-string';
import * as redux from 'redux';

import { mapStateToProps, mapDispatchToProps, TraceDiffImpl } from './TraceDiff';
import TraceDiffHeader from './TraceDiffHeader';
import { actions as diffActions } from './duck';
import * as TraceDiffUrl from './url';
import * as jaegerApiActions from '../../actions/jaeger-api';
import { fetchedState, TOP_NAV_HEIGHT } from '../../constants';

describe('TraceDiff', () => {
  const defaultA = 'trace-id-a';
  const defaultB = 'trace-id-b';
  const defaultCohortIds = ['trace-id-cohort-0', 'trace-id-cohort-1', 'trace-id-cohort-2'];
  const defaultCohort = [defaultA, defaultB, ...defaultCohortIds];
  const fetchMultipleTracesMock = jest.fn();
  const forceStateMock = jest.fn();
  const historyPushMock = jest.fn();
  const defaultProps = {
    a: defaultA,
    b: defaultB,
    cohort: defaultCohort,
    fetchMultipleTraces: fetchMultipleTracesMock,
    forceState: forceStateMock,
    history: {
      push: historyPushMock,
    },
    tracesData: new Map(defaultCohort.map(id => [id, { id, state: fetchedState.DONE }])),
    traceDiffState: {
      a: defaultA,
      b: defaultB,
      cohort: defaultCohort,
    },
  };
  const newAValue = 'newAValue';
  const newBValue = 'newBValue';
  const nonDefaultCohortId = 'non-default-cohort-id';
  const getUrlSpyMockReturnValue = 'getUrlSpyMockReturnValue';
  let getUrlSpy;
  let wrapper;

  beforeAll(() => {
    getUrlSpy = jest.spyOn(TraceDiffUrl, 'getUrl').mockReturnValue(getUrlSpyMockReturnValue);
  });

  beforeEach(() => {
    fetchMultipleTracesMock.mockClear();
    forceStateMock.mockClear();
    getUrlSpy.mockClear();
    historyPushMock.mockClear();
    wrapper = shallow(<TraceDiffImpl {...defaultProps} />);
  });

  describe('syncStates', () => {
    it('forces state if a is inconsistent between url and reduxState', () => {
      wrapper.setProps({ a: newAValue });
      expect(forceStateMock).toHaveBeenLastCalledWith({
        a: newAValue,
        b: defaultProps.b,
        cohort: defaultProps.cohort,
      });
    });

    it('forces state if b is inconsistent between url and reduxState', () => {
      wrapper.setProps({ b: newBValue });
      expect(forceStateMock).toHaveBeenLastCalledWith({
        a: defaultProps.a,
        b: newBValue,
        cohort: defaultProps.cohort,
      });
    });

    it('forces state if cohort size has changed', () => {
      const newCohort = [...defaultProps.cohort, nonDefaultCohortId];
      wrapper.setProps({ cohort: newCohort });
      expect(forceStateMock).toHaveBeenLastCalledWith({
        a: defaultProps.a,
        b: defaultProps.b,
        cohort: newCohort,
      });

      wrapper.setProps({
        cohort: defaultProps.cohort,
        traceDiffState: { ...defaultProps.traceDiffState, cohort: null },
      });
      expect(forceStateMock).toHaveBeenLastCalledWith({
        a: defaultProps.a,
        b: defaultProps.b,
        cohort: defaultProps.cohort,
      });
    });

    it('forces state if cohort entry has changed', () => {
      const newCohort = [...defaultProps.cohort.slice(1), nonDefaultCohortId];
      wrapper.setProps({ cohort: newCohort });
      expect(forceStateMock).toHaveBeenLastCalledWith({
        a: defaultProps.a,
        b: defaultProps.b,
        cohort: newCohort,
      });
    });

    it('does not force state if cohorts have same values in differing orders', () => {
      wrapper.setProps({
        traceDiffState: {
          ...defaultProps.traceDiffState,
          cohort: defaultProps.traceDiffState.cohort.slice().reverse(),
        },
      });
      expect(forceStateMock).not.toHaveBeenCalled();
    });
  });

  it('requests traces lacking a state', () => {
    const newId0 = 'new-id-0';
    const newId1 = 'new-id-1';
    expect(fetchMultipleTracesMock).toHaveBeenCalledTimes(0);
    wrapper.setProps({ cohort: [...defaultProps.cohort, newId0, newId1] });
    expect(fetchMultipleTracesMock).toHaveBeenCalledWith([newId0, newId1]);
    expect(fetchMultipleTracesMock).toHaveBeenCalledTimes(1);
  });

  it('does not request traces if all traces have a state', () => {
    const newId0 = 'new-id-0';
    const newId1 = 'new-id-1';
    expect(fetchMultipleTracesMock).toHaveBeenCalledTimes(0);
    const cohort = [...defaultProps.cohort, newId0, newId1];
    const tracesData = new Map(defaultProps.tracesData);
    tracesData.set(newId0, { id: newId0, state: fetchedState.ERROR });
    tracesData.set(newId1, { id: newId0, state: fetchedState.LOADING });
    wrapper.setProps({ cohort, tracesData });
    expect(fetchMultipleTracesMock).not.toHaveBeenCalled();
  });

  it('updates url when TraceDiffHeader sets a or b', () => {
    wrapper.find(TraceDiffHeader).prop('diffSetA')(newAValue);
    expect(getUrlSpy).toHaveBeenLastCalledWith({
      a: newAValue.toLowerCase(),
      b: defaultProps.b,
      cohort: defaultProps.cohort,
    });

    wrapper.find(TraceDiffHeader).prop('diffSetB')(newBValue);
    expect(getUrlSpy).toHaveBeenLastCalledWith({
      a: defaultProps.a,
      b: newBValue.toLowerCase(),
      cohort: defaultProps.cohort,
    });

    wrapper.find(TraceDiffHeader).prop('diffSetA')('');
    expect(getUrlSpy).toHaveBeenLastCalledWith({
      a: defaultProps.a,
      b: defaultProps.b,
      cohort: defaultProps.cohort,
    });

    wrapper.find(TraceDiffHeader).prop('diffSetB')('');
    expect(getUrlSpy).toHaveBeenLastCalledWith({
      a: defaultProps.a,
      b: defaultProps.b,
      cohort: defaultProps.cohort,
    });

    expect(historyPushMock).toHaveBeenCalledTimes(4);
  });

  describe('render', () => {
    it('renders as expected', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('handles a and b not in props.tracesData', () => {
      const tracesData = new Map(defaultProps.tracesData);
      tracesData.delete(defaultA);
      tracesData.delete(defaultB);
      wrapper.setProps({ tracesData });
      expect(wrapper.find(TraceDiffHeader).props()).toEqual(
        expect.objectContaining({
          a: { id: defaultA },
          b: { id: defaultB },
        })
      );
    });

    it('handles absent a and b', () => {
      wrapper.setProps({ a: null, b: null });
      expect(wrapper.find(TraceDiffHeader).props()).toEqual(expect.objectContaining({ a: null, b: null }));
    });
  });

  describe('TraceDiff--graphWrapper top offset', () => {
    const arbitraryHeight = TOP_NAV_HEIGHT * 2;

    it('initializes as TOP_NAV_HEIGHT', () => {
      expect(wrapper.state().graphTopOffset).toBe(TOP_NAV_HEIGHT);
    });

    it('defaults to TOP_NAV_HEIGHT', () => {
      wrapper.setState({ graphTopOffset: arbitraryHeight });
      wrapper.instance().headerWrapperRef(null);
      expect(wrapper.state().graphTopOffset).toBe(TOP_NAV_HEIGHT);
    });

    it('adjusts TraceDiff--graphWrapper top offset based on TraceDiffHeader height', () => {
      wrapper.instance().headerWrapperRef({ clientHeight: arbitraryHeight });
      expect(wrapper.state().graphTopOffset).toBe(TOP_NAV_HEIGHT + arbitraryHeight);
    });
  });

  describe('mapStateToProps', () => {
    const getOwnProps = ({ a = defaultA, b = defaultB } = {}) => ({
      match: {
        params: {
          a,
          b,
        },
      },
    });
    const makeTestReduxState = ({ cohortIds = defaultCohortIds } = {}) => ({
      router: {
        location: {
          search: queryString.stringify({ cohort: cohortIds }),
        },
      },
      trace: {
        traces: cohortIds.reduce((traces, id) => ({ ...traces, [id]: { id, state: fetchedState.DONE } }), {}),
      },
      traceDiff: {
        a: 'trace-diff-a',
        b: 'trace-diff-b',
      },
    });

    it('gets a and b from ownProps', () => {
      expect(mapStateToProps(makeTestReduxState(), getOwnProps())).toEqual(
        expect.objectContaining({
          a: defaultA,
          b: defaultB,
        })
      );
    });

    it('defaults cohort to empty array if a, b, and cohort are not available', () => {
      expect(
        mapStateToProps(makeTestReduxState({ cohortIds: [] }), getOwnProps({ a: null, b: null })).cohort
      ).toEqual([]);
    });

    it('gets cohort from ownProps and state.router.location.search', () => {
      expect(mapStateToProps(makeTestReduxState(), getOwnProps()).cohort).toEqual([
        defaultA,
        defaultB,
        ...defaultCohortIds,
      ]);
    });

    it('filters falsy values from cohort', () => {
      expect(mapStateToProps(makeTestReduxState(), getOwnProps({ a: null })).cohort).toEqual([
        defaultB,
        ...defaultCohortIds,
      ]);

      expect(mapStateToProps(makeTestReduxState(), getOwnProps({ b: null })).cohort).toEqual([
        defaultA,
        ...defaultCohortIds,
      ]);

      expect(
        mapStateToProps(
          makeTestReduxState({ cohortIds: [...defaultCohortIds, '', nonDefaultCohortId] }),
          getOwnProps()
        ).cohort
      ).toEqual([defaultA, defaultB, ...defaultCohortIds, nonDefaultCohortId]);
    });

    it('filters redundant values from cohort', () => {
      expect(
        mapStateToProps(
          makeTestReduxState({ cohortIds: [...defaultCohortIds, nonDefaultCohortId] }),
          getOwnProps({ a: nonDefaultCohortId })
        ).cohort
      ).toEqual([nonDefaultCohortId, defaultB, ...defaultCohortIds]);

      expect(
        mapStateToProps(
          makeTestReduxState({ cohortIds: [...defaultCohortIds, nonDefaultCohortId] }),
          getOwnProps({ b: nonDefaultCohortId })
        ).cohort
      ).toEqual([defaultA, nonDefaultCohortId, ...defaultCohortIds]);

      expect(
        mapStateToProps(
          makeTestReduxState({ cohortIds: [...defaultCohortIds, nonDefaultCohortId, nonDefaultCohortId] }),
          getOwnProps()
        ).cohort
      ).toEqual([defaultA, defaultB, ...defaultCohortIds, nonDefaultCohortId]);
    });

    // This test may false negative if previous tests are failing
    it('builds tracesData Map from cohort and state.trace.traces', () => {
      const {
        tracesData,
        cohort: { length: expectedSize },
      } = mapStateToProps(makeTestReduxState(), getOwnProps());
      defaultCohortIds.forEach(id => {
        expect(tracesData.get(id)).toEqual({
          id,
          state: fetchedState.DONE,
        });
      });
      expect(tracesData.get(defaultA)).toEqual({
        id: defaultA,
        state: null,
      });
      expect(tracesData.get(defaultB)).toEqual({
        id: defaultB,
        state: null,
      });
      expect(tracesData.size).toBe(expectedSize);
    });

    it('includes state.traceDiff as traceDiffState', () => {
      const testReduxState = makeTestReduxState();
      const { traceDiffState } = mapStateToProps(testReduxState, getOwnProps());
      expect(traceDiffState).toBe(testReduxState.traceDiff);
    });
  });

  describe('mapDispatchToProps', () => {
    let bindActionCreatorsSpy;

    beforeAll(() => {
      bindActionCreatorsSpy = jest.spyOn(redux, 'bindActionCreators').mockImplementation(actions => {
        if (actions === jaegerApiActions) {
          return { fetchMultipleTraces: fetchMultipleTracesMock };
        }
        if (actions === diffActions) {
          return { forceState: forceStateMock };
        }
        return {};
      });
    });

    afterAll(() => {
      bindActionCreatorsSpy.mockRestore();
    });

    it('correctly binds actions to dispatch', () => {
      const dispatchMock = () => {};
      const result = mapDispatchToProps(dispatchMock);
      expect(result.fetchMultipleTraces).toBe(fetchMultipleTracesMock);
      expect(result.forceState).toBe(forceStateMock);
      expect(bindActionCreatorsSpy.mock.calls[0][1]).toBe(dispatchMock);
    });
  });
});
