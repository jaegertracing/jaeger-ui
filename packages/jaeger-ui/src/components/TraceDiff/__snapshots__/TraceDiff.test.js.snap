// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TraceDiff render renders as expected 1`] = `
<Fragment>
  <div
    key="header"
  >
    <TraceDiffHeader
      a={
        Object {
          "id": "trace-id-a",
          "state": "FETCH_DONE",
        }
      }
      b={
        Object {
          "id": "trace-id-b",
          "state": "FETCH_DONE",
        }
      }
      cohort={
        Array [
          Object {
            "id": "trace-id-a",
            "state": "FETCH_DONE",
          },
          Object {
            "id": "trace-id-b",
            "state": "FETCH_DONE",
          },
          Object {
            "id": "trace-id-cohort-0",
            "state": "FETCH_DONE",
          },
          Object {
            "id": "trace-id-cohort-1",
            "state": "FETCH_DONE",
          },
          Object {
            "id": "trace-id-cohort-2",
            "state": "FETCH_DONE",
          },
        ]
      }
      diffSetA={[Function]}
      diffSetB={[Function]}
      key="header"
    />
  </div>
  <div
    className="TraceDiff--graphWrapper"
    key="graph"
    style={
      Object {
        "top": 46,
      }
    }
  >
    <Connect(UnconnectedTraceDiffGraph)
      a={
        Object {
          "id": "trace-id-a",
          "state": "FETCH_DONE",
        }
      }
      b={
        Object {
          "id": "trace-id-b",
          "state": "FETCH_DONE",
        }
      }
    />
  </div>
</Fragment>
`;

exports[`TraceDiff render renders as expected 2`] = `
<Fragment>
  <div
    key="header"
  >
    <TraceDiffHeader
      a={
        Object {
          "id": "trace-id-a",
          "state": "FETCH_DONE",
        }
      }
      b={
        Object {
          "id": "trace-id-b",
          "state": "FETCH_DONE",
        }
      }
      cohort={
        Array [
          Object {
            "id": "trace-id-a",
            "state": "FETCH_DONE",
          },
          Object {
            "id": "trace-id-b",
            "state": "FETCH_DONE",
          },
          Object {
            "id": "trace-id-cohort-0",
            "state": "FETCH_DONE",
          },
          Object {
            "id": "trace-id-cohort-1",
            "state": "FETCH_DONE",
          },
          Object {
            "id": "trace-id-cohort-2",
            "state": "FETCH_DONE",
          },
        ]
      }
      diffSetA={[Function]}
      diffSetB={[Function]}
      key="header"
    />
  </div>
  <div
    className="TraceDiff--graphWrapper"
    key="graph"
    style={
      Object {
        "top": 46,
      }
    }
  >
    <Connect(UnconnectedTraceDiffGraph)
      a={
        Object {
          "id": "trace-id-a",
          "state": "FETCH_DONE",
        }
      }
      b={
        Object {
          "id": "trace-id-b",
          "state": "FETCH_DONE",
        }
      }
    />
  </div>
</Fragment>
`;
